src/core/base.ts

```diff
@@ -34,35 +37,73 @@ export function createPropertyBinding(target: any, key: string): PropertyDescrip
         },
         set(value: any) {
             this[`_${key}`] = value;
-            this.render();
+            this.renderCallback();
         }
     }
 }
 
-@Component('simple-element')
 export class SimpleComponent extends HTMLElement {
 
-    @property()
-    hello: string = 'abc';
+    shadow: ShadowRoot;
+    connected: boolean = false;
+    isSimpleComponent = true;
 
     constructor() {
         super();
-        this.constructor[observedAttributes].forEach((prop: string) => {
-            this.setAttribute(prop, this[prop]);
-        });
+        this.shadow = this.attachShadow({mode: 'open'});
     }
 
     protected connectedCallback(): void {
-        this.innerHTML = `<p>I\'m here!<div>Attribute value: ${this.hello}</div></p>`
+        this.connected = true;
+        this.constructor[observedAttributes].forEach((prop: string) => this.setAttribute(prop, this[toCamelCase(prop)]));
+        this.onInit();
+        this.renderCallback();
+    }
+
+    private attributeChangedCallback(key, oldValue, newValue) {
+        const prop: string = toCamelCase(key);
+        if (newValue !== oldValue && this[bindings].includes(prop)) {
+            this[prop] = newValue;
+            const change = {
+                [prop]: {oldValue, newValue}
+            };
+            this.onChange(change);
+        }
     }
 
-    protected attributeChangedCallback(name, oldValue, newValue) {
-        if (newValue !== oldValue) {
-            this[name] = newValue;
+    protected renderCallback(): void {
+        if (this.connected) {
+            this.beforeRender();
+            this.render();
+            this.afterRender();
         }
     }
 
     protected render(): void {
```

Comment:
Shouldn't `render`, `beforeRender` and `afterRender` be private?

Reply:
No. It's part of lifecycle